### PR TITLE
Type conversion

### PR DESCRIPTION
**Pull request by gabi-250**
This a bit large, so it might be easier to review each commit individually.

* `store_fn_arg` from `ArgTypeMethods` stores a function argument of type `ArgType` into a specified destination (a `PlaceRef`). This commit handles some of the possible `PassMode`s of the argument.
* `new_fn_type` from `AbiMethods` returns the ironox function type for a given `ty::FnSig`. To simplify the implementation (and to be consistent with the way the LLVM backend achieves this), I defined an `FnTypeExt` trait. This trait is implemented for `FnType`s. The functions in `FnTypeExt` make it easy to
create a `FnType` from a `ty::FnSig`, and to convert the `FnType` to an ironox function type.

* Implement `type_named_struct` and `set_struct_body` to create struct types.
* Implement several `BaseTypeMethods`: `type_void`,  `type_struct`, `type_array`, `type_ptr_to`. These all return an ironox `Type`.
* Implement `LayoutIronOxExt` for `TyLayout`.

### CODE REVIEW

File: src/librustc_codegen_ironox/type_.rs [hunk 1]
```diff
@@ -46,25 +47,46 @@ pub enum LLType {
...
 }
 
 impl CodegenCx<'ll, 'tcx> {
-    crate fn type_named_struct(&self, name: &str) -> &Type {
-        unimplemented!("type_named_struct");
+    crate fn type_named_struct(&self, name: &str) -> Type {
+        let mut borrowed_types = self.types.borrow_mut();
+        let struct_type = LLType::StructType {
+            name: Some(name.to_string()),
+            members: vec![],
```
Comment (vext01) [inline src/librustc_codegen_ironox/type_.rs]: So we start with no members, and then fill them in with the function below?

File: src/librustc_codegen_ironox/type_.rs [hunk 2]
```diff
@@ -46,25 +47,46 @@ pub enum LLType {
...
+        members: Vec<Type>
+    },
+    Void,
 }
 
 impl CodegenCx<'ll, 'tcx> {
-    crate fn type_named_struct(&self, name: &str) -> &Type {
-        unimplemented!("type_named_struct");
+    crate fn type_named_struct(&self, name: &str) -> Type {
+        let mut borrowed_types = self.types.borrow_mut();
```
Comment (vext01) [inline src/librustc_codegen_ironox/type_.rs]: Is `borrowed_*` something you saw in the LLVM backend? If not, I'd kill the `borrowed_` prefix.

File: src/librustc_codegen_ironox/type_.rs [hunk 3]
```diff
@@ -46,25 +47,46 @@ pub enum LLType {
...
+            return;
+        }
+        bug!("expected LLType::StructType, found {:?}", borrowed_types[ty]);
     }
 }
 
 impl BaseTypeMethods<'tcx> for CodegenCx<'ll, 'tcx> {
     fn type_void(&self) -> Type {
-        unimplemented!("type_void");
+        let mut borrowed_types = self.types.borrow_mut();
```
Comment (vext01) [inline src/librustc_codegen_ironox/type_.rs]: So can we end up pushing duplicates if this method is called more than once?

File: src/librustc_codegen_ironox/type_.rs [hunk 4]
```diff
@@ -158,11 +180,23 @@ impl BaseTypeMethods<'tcx> for CodegenCx<'ll, 'tcx> {
         els: &[Type],
         packed: bool
     ) -> Type {
-        unimplemented!("type_struct");
+        let mut borrowed_types = self.types.borrow_mut();
+        let mut members = els.to_vec();
+        let struct_type = LLType::StructType {
+            name: None,
```
Comment (vext01) [inline src/librustc_codegen_ironox/type_.rs]: I guess I was expecting a named struct version too?

File: src/librustc_codegen_ironox/type_.rs [hunk 5]
```diff
@@ -200,7 +236,13 @@ impl BaseTypeMethods<'tcx> for CodegenCx<'ll, 'tcx> {
     fn val_ty(&self, v: Value) -> Type {
         // FIXME: return the real type
         let mut borrowed_types = self.types.borrow_mut();
-        borrowed_types.push(LLType::None);
+        match v {
+            Value::ConstStruct(_) => {
+                // FIXME: implement
```
Comment (vext01) [inline src/librustc_codegen_ironox/type_.rs]: Can this be unimplemented!? To be consistent?

File: src/librustc_codegen_ironox/type_of.rs [hunk 6]
```diff
@@ -10,17 +17,141 @@ pub trait LayoutIronOxExt<'tcx> {
...
 }
 
+/// Return the IronOx `Type`s of the fields of the specified `layout`.
+fn struct_field_types(
+    cx: &CodegenCx<'_, 'tcx>,
+    layout: TyLayout<'tcx>) -> Vec<Type> {
+    let mut fields = vec![];
+    for i in layout.fields.index_by_increasing_offset() {
+        let field = layout.field(cx, i);
+        // FIXME: handle field alignment
```
Comment (vext01) [inline src/librustc_codegen_ironox/type_of.rs]: Scary :)

File: src/librustc_codegen_ironox/type_of.rs [hunk 7]
```diff
@@ -10,17 +17,141 @@ pub trait LayoutIronOxExt<'tcx> {
...
+        fields.push(field.ironox_type(cx));
+    }
+    fields
+}
+
 impl LayoutIronOxExt<'tcx> for TyLayout<'tcx> {
     fn ironox_type(&self, cx: &CodegenCx<'_, 'tcx>) -> Type {
-        unimplemented!("ironox_type");
+        if let layout::Abi::Scalar(ref scalar) = self.abi {
```
Comment (vext01) [inline src/librustc_codegen_ironox/type_of.rs]: Can you explain the purpose of the `ironox_type` method here?

File: src/librustc_codegen_ironox/type_.rs (hunk 1 above)
Comment (gabi-250) [reply src/librustc_codegen_ironox/type_.rs]: Yep. That seems to be the way the backend-agnostic code puts together a struct.

File: src/librustc_codegen_ironox/type_.rs (hunk 3 above)
Comment (gabi-250) [reply src/librustc_codegen_ironox/type_.rs]: Afraid so... I need to come up with a better way of passing types around.

File: src/librustc_codegen_ironox/type_.rs (hunk 4 above)
Comment (gabi-250) [reply src/librustc_codegen_ironox/type_.rs]: There is one: https://github.com/gabi-250/rust/blob/508c80943387801b1f3d2701b8562652d3f6ab1e/src/librustc_codegen_ironox/type_.rs#L65

File: src/librustc_codegen_ironox/type_of.rs (hunk 7 above)
Comment (gabi-250) [reply src/librustc_codegen_ironox/type_of.rs]: It returns the `Type` of a Rust type. Used here:

https://github.com/gabi-250/rust/blob/508c80943387801b1f3d2701b8562652d3f6ab1e/src/librustc_codegen_ironox/type_.rs#L255-L257

and here:

https://github.com/gabi-250/rust/blob/508c80943387801b1f3d2701b8562652d3f6ab1e/src/librustc_codegen_ironox/type_.rs#L259-L261

`Type` is a type I defined for my backend. All codegen backends have to define their own `Type` (they don't operate directly on Rust types such as `TyLayout`).

Basically, it transofrms a `TyLayout` (Rust type) into a `Type` (backend-specific type).

File: src/librustc_codegen_ironox/type_.rs (hunk 2 above)
Comment (gabi-250) [reply src/librustc_codegen_ironox/type_.rs]: Done.

File: src/librustc_codegen_ironox/type_.rs (hunk 5 above)
Comment (gabi-250) [reply src/librustc_codegen_ironox/type_.rs]: Done.

File: src/librustc_codegen_ironox/type_.rs (hunk 3 above)
Comment (vext01) [reply src/librustc_codegen_ironox/type_.rs]: Could you use a `HashSet`? If you rely on the ordering of the collection, then no.

File: src/librustc_codegen_ironox/type_.rs (hunk 4 above)
Comment (vext01) [reply src/librustc_codegen_ironox/type_.rs]: OK

File: src/librustc_codegen_ironox/type_.rs (hunk 5 above)
Comment (vext01) [reply src/librustc_codegen_ironox/type_.rs]: Thanks

**Comment by vext01** [comment]
This looks fine to me. Shall we wait to see if Laurie has any comments?

File: src/librustc_codegen_ironox/abi.rs [hunk 8]
```diff
@@ -68,14 +105,74 @@ impl AbiMethods<'tcx> for CodegenCx<'ll, 'tcx> {
...
+        // Create the return type.
+        let ret_ty = match self.ret.mode {
+            PassMode::Ignore => cx.type_void(),
+            PassMode::Direct(_) => {
+                self.ret.layout.immediate_ironox_type(cx)
+            },
+            mode => unimplemented!("{:?}", mode)
+        };
+        // Create the types of the arguments.
+        let mut arg_tys = vec![];
```
Comment (ltratt) [inline src/librustc_codegen_ironox/abi.rs]: Not a big deal, but I'd probably do `Vec::with_capacity(self.args.len())`.

File: src/librustc_codegen_ironox/type_.rs [hunk 9]
```diff
@@ -158,11 +180,23 @@ impl BaseTypeMethods<'tcx> for CodegenCx<'ll, 'tcx> {
...
+            members,
+        };
+        types.push(struct_type);
+        types.len() - 1
     }
 
     fn type_array(&self, ty: Type, len: u64) -> Type {
-        unimplemented!("type_array");
+        let mut types = self.types.borrow_mut();
+        let array_type = LLType::Array {
```
Comment (ltratt) [inline src/librustc_codegen_ironox/type_.rs]: For consistency with the code below, I'd probably do `types.push(LLType::Array { ... })`.

File: src/librustc_codegen_ironox/type_.rs [hunk 10]
```diff
@@ -199,9 +235,14 @@ impl BaseTypeMethods<'tcx> for CodegenCx<'ll, 'tcx> {
 
     fn val_ty(&self, v: Value) -> Type {
         // FIXME: return the real type
-        let mut borrowed_types = self.types.borrow_mut();
-        borrowed_types.push(LLType::None);
-        borrowed_types.len() - 1
+        let mut types = self.types.borrow_mut();
+        match v {
```
Comment (ltratt) [inline src/librustc_codegen_ironox/type_.rs]: This might be easier as `if let Value::ConstStruct(_) = v { unimplemented(...) }`.

**Comment by ltratt** [comment]
I am even more out of my depth on this than Edd, so my comments are very shallow.

**Comment by gabi-250** [comment]
Thank you for your comments! 

I think I've addressed everything, so this should be ready for a re-review.

File: src/librustc_codegen_ironox/type_.rs (hunk 3 above)
Comment (ltratt) [reply src/librustc_codegen_ironox/type_.rs]: This sounds like something that should probably be addressed in this PR if possible. Is it difficult?

Comment (gabi-250) [reply src/librustc_codegen_ironox/type_.rs]: It shouldn't be *too* hard. I'll see what I can do.

**Comment by gabi-250** [comment]
I added a `type_cache` to `CodegenCx`. If an `LLType` is already in `type_cache`, its index in the `types` vector is returned. If not, it is added to `types`, and to the `type_cache`. Ready for re-review.

**Comment by vext01** [comment]
This looks fine to me. If Laurie is good with it, I think we can then squash.

File: src/librustc_codegen_ironox/type_.rs [hunk 11]
```diff
@@ -62,14 +62,32 @@ pub enum LLType {
...
+    ///
+    /// If the type already exists in the type cache, its index in the type
+    /// vector is returned. Otherwise, it is added to the type vector, and to
+    /// the type cache.
+    fn add_type(&self, ll_type: LLType) -> Type {
         let mut types = self.types.borrow_mut();
+        // If ll_type is not in types, it will be inserted at the end of the
+        // types vector.
+        let ty_idx = types.len();
+        // FIXME: don't always clone ll_type.
```
Comment (ltratt) [inline src/librustc_codegen_ironox/type_.rs]: Is this a performance issue or a semantic issue?

Comment (gabi-250) [reply src/librustc_codegen_ironox/type_.rs]: It's a performance issue.

Comment (ltratt) [reply src/librustc_codegen_ironox/type_.rs]: OK, then we can ignore that for the time being.

**Comment by ltratt** [comment]
Please squash.

**Comment by gabi-250** [comment]
Squashed!
